Frame callback implementation & tests

denniszollo opened the pull request:
This PR adds a new type of callback to the c SBP Library to handle the entire SBP frame.  It leaves the external API unchanged except for these additions.

This new frame callback feature will allow for easier raw logging in applications including Piksi Multi's standalone logger.  It also allows adding a frame callback of type SBP_MSG_ALL to fire a callback for any msg_type.  A convenient helper function is added to the API for registering a callback for every msg_type.  Implementation details were discussed internally quite a bit and I think this represents the consensus.  

I've added unit tests for the new functionality and some missing unit tests like testing the max_size of SBP messages.

benjaminaltieri commented on c/src/sbp.c:
```diff
@@ -381,8 +457,15 @@ s8 sbp_process(sbp_state_t *s, s32 (*read)(u8 *buff, u32 n, void *context))
+         * In future, it may be wise to add new return codes for the truth
+         * table of both process function return codes. For now, we return only
+         * sbp_process_payload's return code for backwards compatibility. */
+        sbp_process_frame(s, s->sender_id, s->msg_type, s->msg_len, s->msg_buff,
```
I would propose we allow `sbp_process_frame` to take an argument such that it calls all or a subset of callback types. this would allow `sbp_process_payload` to be a specialized call to `sbp_process_frame` where only `SBP_TYPE_CALLBACK` nodes are executed (similar to how you are extending the callback register functions

benjaminaltieri replied:
this would reduce code duplication between the two methods as well as avoid having to traverse the list twice.

benjaminaltieri commented on c/test/check_sbp.c:
```diff
@@ -83,11 +83,21 @@ u8 last_msg[256];
 void logging_reset(void)
 {
   n_callbacks_logged = 0;
+  n_frame_callbacks_logged = 0;
```
should the `last_*` variables get reset too?

denniszollo replied:
yes, that would be better. updated.

benjaminaltieri commented on c/test/check_sbp.c:
```diff
@@ -461,6 +815,10 @@ Suite* sbp_suite(void)
   tcase_add_test(tc_core, test_callbacks);
   tcase_add_test(tc_core, test_sbp_send_message);
   tcase_add_test(tc_core, test_sbp_process);
+  tcase_add_test(tc_core, test_sbp_frame);
```
adding tests too! this is great, much good karma to you 🕉

benjaminaltieri commented on c/include/libsbp/sbp.h:
```diff
@@ -43,19 +43,42 @@ extern "C" {
+ */
+enum sbp_cb_type{
+  SBP_TYPE_CALLBACK,
+  SBP_FRAME_CALLBACK
```
in true C style, we could turn this into a bit field to allow the `sbp_process_frame` function to be super flexible, just throw in a bitmask to select the callback styles you'd like to service. as easy as:
```
    if ((node->cb_type & callback_select_mask) &&
        ((node->msg_type == msg_type) || (node->msg_type == SBP_MSG_ALL))) {
```

benjaminaltieri commented on c/src/sbp.c:
```diff
@@ -323,6 +384,12 @@ s8 sbp_process(sbp_state_t *s, s32 (*read)(u8 *buff, u32 n, void *context))
         s->n_read = 0;
+        /* set frame_buff to 0 after each preamble.
+         * Note, library functions are not used to avoid more dependencies. */
+        for (int i = 0; i < 255; i++) {
```
`255` doesn't seem right here. since it's a static buffer we can use `sizeof(s->frame_buff)`

benjaminaltieri commented:
Put these changes together to help illustrate the ideas: https://github.com/swift-nav/libsbp/tree/benaltieri/frame_callback_ideas

denniszollo commented:
Ready for review.

denniszollo commented on c/src/sbp.c:
```diff
@@ -315,76 +401,89 @@ s8 sbp_process(sbp_state_t *s, s32 (*read)(u8 *buff, u32 n, void *context))
-      crc = crc16_ccitt((u8*)&(s->sender_id), 2, crc);
-      crc = crc16_ccitt(&(s->msg_len), 1, crc);
-      crc = crc16_ccitt(s->msg_buff, s->msg_len, crc);
+      s->crc = (s->frame_buff[SBP_FRAME_OFFSET_CRC(s->msg_len)] +
```
You suggest a macro like  
s->crc = SBP_FRAME_TO_U16(buff, start_offset), correct?

denniszollo replied:
How does this look: 
![image](https://user-images.githubusercontent.com/11276774/64885395-ede0c280-d618-11e9-88a4-fe1c1707d5f0.png)

denniszollo replied:
I decided to use a function to allow the compiler to help a user.  With a macro, I couldn't figure out how to make it safe and have informative error messages.

silverjam replied:
👍

denniszollo commented on c/src/sbp.c:
```diff
@@ -147,20 +147,27 @@
+ * \param s        Pointer to sbp_state
+ * \param msg_type message type on which to fire callback
+ *                 SBP_MSG_ALL will fire for every message only on frame
+ *                 callbacks.
```
Will shortly push an updated comment to make more explicit and clear.

denniszollo commented on c/src/sbp.c:
```diff
@@ -266,6 +327,31 @@ void sbp_state_set_io_context(sbp_state_t *s, void *context)
   s->io_context = context;
 }
 
+/** Helper to fill frame buffer in context from read function and increment read count
```
I added a full docstring to this function, which is worth it for our published doxygen (https://swift-nav.github.io/libsbp/c/build/docs/html/)

denniszollo commented on c/src/sbp.c:
```diff
@@ -266,6 +327,31 @@ void sbp_state_set_io_context(sbp_state_t *s, void *context)
+    return SBP_OK;
+}
+
+/** Reset frame buffer in context to zeros and reset length
```
I think we meant on sbp_state.  Updated in forthcoming commit.

denniszollo commented on c/src/sbp.c:
```diff
@@ -415,18 +514,62 @@ s8 sbp_process(sbp_state_t *s, s32 (*read)(u8 *buff, u32 n, void *context))
+
+/** Directly process an SBP frame.
+ * Use this function to directly process the entire SBP frame after
+ * it is succesfully deframed and the CRC has passed. It will fire ONLY
```
This is a stale comment indeed.  I worked to improve on forthcoming push to this branch.

denniszollo commented on c/src/sbp.c:
```diff
@@ -415,18 +514,62 @@ s8 sbp_process(sbp_state_t *s, s32 (*read)(u8 *buff, u32 n, void *context))
+ * it is succesfully deframed and the CRC has passed. It will fire ONLY
+ * the frame callbacks registered for the message.
+ *
+ * \param s State structure
```
Done.  It would be nice if clang-format could handle doxygen.

denniszollo commented:
Doxygen looks pretty good:
![image](https://user-images.githubusercontent.com/11276774/64887092-a2c8ae80-d61c-11e9-8878-4bf71b942b14.png)
![image](https://user-images.githubusercontent.com/11276774/64887138-c7248b00-d61c-11e9-87aa-ca6fdff0b025.png)

denniszollo commented on c/src/sbp.c:
```diff
@@ -315,76 +401,89 @@ s8 sbp_process(sbp_state_t *s, s32 (*read)(u8 *buff, u32 n, void *context))
+      if (expected_frame_len != s->frame_len) {
+        return SBP_READ_TOO_MANY_ERROR;
+      }
       /* Swap bytes to little endian. */
```
Per this good feedback I removed the bogus comment and created a helper function to cast to u8s to a u16.  I would love to generate real parsing functions that are platform independent for each field for SBP with the generator and hope to accomplish that.

silverjam replied:
Big endian seems pretty uncommon lately, so I think it's a "nice to have", we have definitely seen (1 or 2) customers with this issue, but it shouldn't be high priority.

denniszollo commented on c/include/libsbp/sbp.h:
```diff
@@ -23,39 +23,91 @@ extern "C" {
+ * to cast the callback function pointer stored within.
+ */
+enum sbp_cb_type {
+  SBP_TYPE_CALLBACK = 0,
```
I am starting to think this should be SBP_PAYLOAD_CALLBACK instead of SBP_TYPE_CALLBACK.  I had named it type callback when it was impossible to register a frame callback on a particular message type.

denniszollo commented:
I've made a few changes. I added a helper function to go from u8[] to u16 and remove confusing endianness comments, renamed SBP_TYPE_CALLBACK to SBP_PAYLOAD_CB, and updated the docs a lot per Jason's suggestions and thoughts.  If approved I will plan to rebase and merge as is with the relatively fine-grained commit history. 

If requested, I am willing and able to squash into three logical commits. 
  1. code changes for frame_callback
  2. comment and doc updates for frame callback
  3. unit test additions for frame_callback

silverjam commented:
@denniszollo I think some squashing is in order, whatever seems appropriate to you is fine.

denniszollo commented:
After working with this new API quite a bit on Friday, it became clear to me that the new return code (SBP_READ_TOO_MANY_ERROR)  could result in a change in behavior depending on how the *read function pointer was implemented, as such Ben and I decided to remove it.  I am going to merge the newly squashed version of this with only this one change.

benjaminaltieri commented:
@denniszollo squashing helped, my only nit would be that the macro commit should be before the commit that uses the macros implemented there. one example is `SBP_MAX_FRAME_LEN` use in `sbp_state_t` my suggestion would be to squash that or just reorder. The reason being is that if someone is looking at the full code in that single functional commit, they will not see the definition of those macros
